### PR TITLE
Service Water Heating module

### PR DESCRIPTION
**Pull request by mdahlhausen**
Pull request overview
---------------------
This PR creates the ServiceWaterHeating module.

It renames and moves methods in `Prototype.ServiceWaterHeating` to the new module, moves methods in `Prototype.ServiceWaterHeating` and `Prototype.Model.swh` to `Standards.ServiceWaterHeating`, then deletes the now empty `Prototype.ServiceWaterHeating` and `Prototype.Model.swh`.

### `create_water_heater.rb`
rename `model_add_water_heater` from `Prototype.ServiceWaterHeating` to `create_water_heater`
rename `model_add_heatpump_water_heater` from `Prototype.ServiceWaterHeating` to `create_heatpump_water_heater`

### `create_water_heating_loop.rb`
rename `model_add_swh_loop` from `Prototype.ServiceWaterHeating` to `create_service_water_heating_loop`
rename `model_add_swh_booster` from `Prototype.ServiceWaterHeating` to `create_booster_water_heating_loop`

### `create_piping_losses.rb`
rename `model_add_piping_losses_to_swh_system` from `Prototype.ServiceWaterHeating` to `create_service_water_heating_piping_losses`

### create_water_use.rb
rename `model_add_swh_end_uses` from `Prototype.ServiceWaterHeating` to `create_water_use`

delete `model_add_booster_swh_end_uses` from `Prototype.ServiceWaterHeating` 
delete `model_attach_water_fixtures_to_spaces` from `Prototype.ServiceWaterHeating` 

### Moved to `Standards.ServiceWaterHeating`:
move `model_add_swh_end_uses_by_space` from `Prototype.ServiceWaterHeating`
move `model_add_swh` from `Prototype.Model.swh`
move `model_add_typical_swh` from `Prototype.Model.swh`
move `model_find_water_heater_capacity_volume_and_parasitic` from `Prototype.Model.swh`

### Changes to other NREL measures
openstudio-load-flexibility-measures-gem\lib\measures\add_hpwh\measure.rb#L517
`std.model_add_heatpump_water_heater` -> `OpenstudioStandards::ServiceWaterHeating.create_heatpump_water_heater`

openstudio-load-flexibility-measures-gem\lib\measures\add_hpwh\measure.rb#L533
`std.model_add_water_heater` -> `OpenstudioStandards::ServiceWaterHeating.create_water_heater`

OpenStudio-measures\lib\measures\multifamily_central_wwhp\measure.rb#L343
`std.model_add_piping_losses_to_swh_system` -> `OpenstudioStandards::ServiceWaterHeating.create_service_water_heating_piping_losses`

### Pull Request Author
 - [x] Method changes or additions
 - [x] Added tests for added methods
 - [x] If methods have been deprecated, update rest of code to use the new methods
 - [x] Documented new methods using [yard syntax](https://rubydoc.info/gems/yard/file/docs/GettingStarted.md)
 - [x] Resolved yard documentation errors for new code (ran `bundle exec rake doc`)
 - [x] Resolved rubocop syntax errors for new code (ran `bundle exec rake rubocop`)
 - [x] All new and existing tests passes

### Review Checklist
 - [x] Perform a code review on GitHub
 - [x] All related changes have been implemented: method additions, changes, tests
 - [ ] Check rubocop errors
 - [ ] Check yard doc errors
 - [ ] If fixing a defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If a new feature, test the new feature and try creative ways to break it
 - [x] CI status: all green or justified

### CODE REVIEW

**Comment by mdahlhausen** [comment]
@ckirney @phylroy this module refactor impacts NECB models. Please review.

File: lib/openstudio-standards/service_water_heating/create_piping_losses.rb [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+      else
+        pipe_construction.setName('Uninsulated copper pipe 0.75in type L')
+        pipe_construction.setLayers([copper_pipe])
+      end
+
+      heat_loss_pipe = OpenStudio::Model::PipeIndoor.new(model)
+      heat_loss_pipe.setName("#{service_water_loop.name} Pipe #{pipe_length_ft.round}ft")
+      heat_loss_pipe.setEnvironmentType('Schedule')
+      # @todoschedule type registry error for this setter
+      # heat_loss_pipe.setAmbientTemperatureSchedule(swh_piping_air_temp_sch)
```
Comment (eringold) [inline lib/openstudio-standards/service_water_heating/create_piping_losses.rb:138]: Looks like this has been fixed: https://github.com/NREL/OpenStudio/issues/4894

File: lib/openstudio-standards/service_water_heating/create_piping_losses.rb [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+      # @todo schedule type registry error for this setter
+      # heat_loss_pipe.setAmbientAirVelocitySchedule(model.alwaysOffDiscreteSchedule)
+      heat_loss_pipe.setPointer(8, swh_piping_air_velocity_sch.handle)
+      heat_loss_pipe.setConstruction(pipe_construction)
+      heat_loss_pipe.setPipeInsideDiameter(OpenStudio.convert(0.785, 'in', 'm').get)
+      heat_loss_pipe.setPipeLength(OpenStudio.convert(pipe_length_ft, 'ft', 'm').get)
+
+      heat_loss_pipe.addToNode(service_water_loop.demandInletNode)
+
+      OpenStudio.logFree(OpenStudio::Info, 'openstudio.model.Model', "Added #{pipe_length_ft.round}ft of #{pipe_construction.name} losing heat to #{air_temperature_f.round}F air to #{service_water_loop.name}.")
```
Comment (eringold) [inline lib/openstudio-standards/service_water_heating/create_piping_losses.rb:149]: Update log messages with the right module name

File: lib/openstudio-standards/service_water_heating/create_water_heater.rb [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+
+      # set water temperature properties
+      water_heater.setDeadbandTemperatureDifference(2.0)
+      water_heater.setDeadbandTemperatureDifference(OpenStudio.convert(3.6, 'R', 'K').get)
+      water_heater.setHeaterControlType('Cycle')
+      water_heater.setOffCycleParasiticHeatFractiontoTank(0.8)
+      water_heater.setIndirectWaterHeatingRecoveryTime(1.5) # 1.5hrs
+
+      # get or create temperature schedule type limits
+      temp_sch_type_limits = OpenstudioStandards::Schedules.create_schedule_type_limits(model,
```
Comment (eringold) [inline lib/openstudio-standards/service_water_heating/create_water_heater.rb:123]: don't need

File: lib/openstudio-standards/service_water_heating/create_water_heater.rb [hunk 4]
```diff
@@ -0,0 +1,571 @@
...
+      if heat_pump_type == 'WrappedCondenser'
+        hpwh.setMinimumInletAirTemperatureforCompressorOperation(hpwh_op_min_temp_c)
+        hpwh.setMaximumInletAirTemperatureforCompressorOperation(hpwh_op_max_temp_c)
+        # set sensor heights
+        if hpwh_vol_gal <= 50.0
+          hpwh.setDeadBandTemperatureDifference(0.5)
+          h_ue = (1 - (3.5 / 12.0)) * tank_height # in the 4th node of the tank (counting from top)
+          h_le = (1 - (10.5 / 12.0)) * tank_height # in the 11th node of the tank (counting from top)
+          h_condtop = (1 - (5.5 / 12.0)) * tank_height # in the 6th node of the tank (counting from top)
+          h_condbot = (1 - (10.99 / 12.0)) * tank_height # in the 11th node of the tank
```
Comment (eringold) [inline lib/openstudio-standards/service_water_heating/create_water_heater.rb:254]: should this be 10.5?

File: lib/openstudio-standards/service_water_heating/create_water_heater.rb [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+      # set heat pump water heater properties
+      hpwh.setFanPlacement('DrawThrough')
+      hpwh.setOnCycleParasiticElectricLoad(0.0)
+      hpwh.setOffCycleParasiticElectricLoad(0.0)
+      hpwh.setParasiticHeatRejectionLocation('Outdoors')
+
+      # set temperature setpoint schedule
+      if service_water_temperature_schedule.nil?
+        # temperature schedule type limits
+        temp_sch_type_limits = OpenstudioStandards::Schedules.create_schedule_type_limits(model,
```
Comment (eringold) [inline lib/openstudio-standards/service_water_heating/create_water_heater.rb:289]: not needed

File: lib/openstudio-standards/service_water_heating/create_water_heater.rb [hunk 6]
```diff
@@ -0,0 +1,571 @@
...
+
+      if water_heater_thermal_zone.nil?
+        # add in schedules for Tamb, RHamb, and the compressor
+        # assume the water heater is indoors at 71.6F / 22C
+        default_water_heater_ambient_temp_sch = OpenstudioStandards::Schedules.create_constant_schedule_ruleset(model,
+                                                                                                                OpenStudio.convert(71.6, 'F', 'C').get,
+                                                                                                                name: 'Water Heater Ambient Temp Schedule 70F',
+                                                                                                                schedule_type_limit: 'Temperature')
+        if temp_sch_type_limits.nil?
+          temp_sch_type_limits = OpenstudioStandards::Schedules.create_schedule_type_limits(model,
```
Comment (eringold) [inline lib/openstudio-standards/service_water_heating/create_water_heater.rb:445]: not needed

**Review by eringold — CHANGES_REQUESTED** [review]

File: lib/openstudio-standards/service_water_heating/create_water_use.rb [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                                                                                                    name: "Fraction Sensible - #{sensible_fraction}",
+                                                                                                    schedule_type_limit: 'Fractional')
+      water_use_latent_frac_sch = OpenstudioStandards::Schedules.create_constant_schedule_ruleset(model,
+                                                                                                  latent_fraction,
+                                                                                                  name: "Fraction Latent - #{latent_fraction}",
+                                                                                                  schedule_type_limit: 'Fractional')
+      water_use_def.setSensibleFractionSchedule(water_use_sensible_frac_sch)
+      water_use_def.setLatentFractionSchedule(water_use_latent_frac_sch)
+      water_use_def.setPeakFlowRate(flow_rate)
+      water_use_def.setName("#{name} #{flow_rate_gpm.round(2)}gpm #{water_use_temperature_f.round}F")
```
[the remark this answers is not in the document]
Comment (mdahlhausen) [reply lib/openstudio-standards/service_water_heating/create_water_use.rb:53]: changed to 1

File: lib/openstudio-standards/service_water_heating/create_water_heating_loop.rb [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+      service_water_loop.setMinimumLoopTemperature(10.0)
+      service_water_loop.setMaximumLoopTemperature(82.2)
+
+      if system_name.nil?
+        system_name = 'Service Water Loop'
+      end
+      service_water_loop.setName(system_name)
+
+      # temperature schedule type limits
+      temp_sch_type_limits = OpenstudioStandards::Schedules.create_schedule_type_limits(model,
```
[the remark this answers is not in the document]
Comment (mdahlhausen) [reply lib/openstudio-standards/service_water_heating/create_water_heating_loop.rb:58]: removed

File: lib/openstudio-standards/service_water_heating/create_water_heating_loop.rb [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+                                                                                          off_cycle_parasitic_fuel_consumption_rate: off_cycle_parasitic_fuel_consumption_rate,
+                                                                                          service_water_temperature: service_water_temperature,
+                                                                                          service_water_temperature_schedule: service_water_temperature_schedule,
+                                                                                          water_heater_thermal_zone: water_heater_thermal_zone,
+                                                                                          service_water_loop: booster_service_water_loop)
+      booster_water_heater.setName("#{water_heater_volume_gal}gal #{water_heater_fuel} Booster Water Heater - #{water_heater_capacity_kbtu_per_hr.round}kBtu/hr")
+      booster_water_heater.setEndUseSubcategory('Booster')
+
+      # Temperature schedule type limits
+      temp_sch_type_limits = OpenstudioStandards::Schedules.create_schedule_type_limits(model,
```
[the remark this answers is not in the document]
Comment (mdahlhausen) [reply lib/openstudio-standards/service_water_heating/create_water_heating_loop.rb:200]: removed

File: lib/openstudio-standards/service_water_heating/create_piping_losses.rb (hunk 2 above)
Comment (mdahlhausen) [reply lib/openstudio-standards/service_water_heating/create_piping_losses.rb:149]: changed logging messages to 'openstudio.standards.ServiceWaterHeating'

File: lib/openstudio-standards/service_water_heating/create_water_heater.rb (hunk 3 above)
Comment (mdahlhausen) [reply lib/openstudio-standards/service_water_heating/create_water_heater.rb:123]: removed

File: lib/openstudio-standards/service_water_heating/create_water_heater.rb (hunk 5 above)
Comment (mdahlhausen) [reply lib/openstudio-standards/service_water_heating/create_water_heater.rb:289]: removed

File: lib/openstudio-standards/service_water_heating/create_water_heater.rb (hunk 6 above)
Comment (mdahlhausen) [reply lib/openstudio-standards/service_water_heating/create_water_heater.rb:445]: removed

File: lib/openstudio-standards/service_water_heating/create_water_use.rb [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+                              sensible_fraction: 0.2,
+                              latent_fraction: 0.05,
+                              service_water_loop: nil,
+                              space: nil)
+      # IP conversions for naming
+      flow_rate_gpm = OpenStudio.convert(flow_rate, 'm^3/s', 'gal/min').get
+      water_use_temperature_f = OpenStudio.convert(water_use_temperature, 'C', 'F').get
+
+      # default name
+      name = 'Main' if name.nil?
```
[the remark this answers is not in the document]
Comment (mdahlhausen) [reply lib/openstudio-standards/service_water_heating/create_water_use.rb:36]: changed to 'Main Water Use'

File: lib/openstudio-standards/service_water_heating/create_piping_losses.rb [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        # get number of water heaters
+        if water_heater.additionalProperties.getFeatureAsInteger('component_quantity').is_initialized
+          comp_qty = water_heater.additionalProperties.getFeatureAsInteger('component_quantity').get
+        else
+          comp_qty = 1
+        end
+
+        if comp_qty > 1
+          OpenStudio.logFree(OpenStudio::Info, 'openstudio.model.Model', "Piping length has been multiplied by #{comp_qty}X because #{water_heater.name} represents #{comp_qty} pieces of equipment.")
+          pipe_length_ft *= comp_qty
```
[the remark this answers is not in the document]
Comment (mdahlhausen) [reply lib/openstudio-standards/service_water_heating/create_piping_losses.rb:74]: I think the idea is they may be in different parts of the building. It's part of the method PNNL came up with for the PRM methodology.

I adjusted it so the user pipe length will stay as provided and not get multiplied if there is more than 1 water heater.

File: lib/openstudio-standards/service_water_heating/create_water_heating_loop.rb [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+                                                                                        name: 'Temperature Schedule Type Limits',
+                                                                                        lower_limit_value: 0.0,
+                                                                                        upper_limit_value: 100.0,
+                                                                                        numeric_type: 'Continuous',
+                                                                                        unit_type: 'Temperature')
+
+      # service water heating loop controls
+      swh_temp_f = OpenStudio.convert(service_water_temperature, 'C', 'F').get
+      swh_delta_t_r = 9.0 # 9F delta-T
```
[the remark this answers is not in the document]
Comment (mdahlhausen) [reply lib/openstudio-standards/service_water_heating/create_water_heating_loop.rb:67]: I'm not sure how important it is. I'm going to keep it the same for now to match existing models. We can change it when we get new validation data on service water systems that shows a different assumption is better.

File: lib/openstudio-standards/service_water_heating/create_water_heating_loop.rb [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+      if site_water_mains.temperatureSchedule.is_initialized
+        water_mains_temp_sch = site_water_mains.temperatureSchedule.get
+      elsif site_water_mains.annualAverageOutdoorAirTemperature.is_initialized
+        mains_src_temp_c = site_water_mains.annualAverageOutdoorAirTemperature.get
+        mains_src.setSourceTemperature(mains_src_temp_c)
+        water_mains_temp_sch = OpenStudio::Model::ScheduleConstant.new(model)
+        water_mains_temp_sch.setName('Booster Water Makeup Temperature')
+        water_mains_temp_sch.setValue(mains_src_temp_c)
+      else # assume 50F
+        mains_src_temp_c = OpenStudio.convert(50.0, 'F', 'C').get
```
[the remark this answers is not in the document]
Comment (mdahlhausen) [reply lib/openstudio-standards/service_water_heating/create_water_heating_loop.rb:278]: I'm not sure; I don't know if the booster water heater is structured correctly with makeup water. That can be a part of future validation work.

File: lib/openstudio-standards/service_water_heating/create_water_heater.rb (hunk 4 above)
Comment (mdahlhausen) [reply lib/openstudio-standards/service_water_heating/create_water_heater.rb:254]: I think it's correct. The comment may have been copied and pasted without being updated. It's from the residential hpxml code. @jmaguire1 I think was the original author of this and may know.

File: lib/openstudio-standards/service_water_heating/create_water_heating_loop.rb [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+        swh_pump = OpenStudio::Model::PumpVariableSpeed.new(model)
+        swh_pump.setName("#{service_water_loop.name} Water Mains Pressure Driven")
+        swh_pump.setPumpControlType('Continuous')
+      end
+      swh_pump.setRatedPumpHead(service_water_pump_head.to_f)
+      swh_pump.setMotorEfficiency(service_water_pump_motor_efficiency)
+      swh_pump.addToNode(service_water_loop.supplyInletNode)
+
+      # add water heater
+      water_heater = OpenstudioStandards::ServiceWaterHeating.create_water_heater(model,
```
[the remark this answers is not in the document]
Comment (mdahlhausen) [reply lib/openstudio-standards/service_water_heating/create_water_heating_loop.rb:107]: Yeah, that would be annoying to have to delete the water heater to use the hpwh. I added an option for 'None', which won't make a water heater, and added an option to use the formal 'HeatPump' water heater model.

**Review by eringold — APPROVED** [review]
Changes look good. Thanks @mdahlhausen!

**Pull request merged**